protocol/consensus-nonces.md

```diff
@@ -0,0 +1,186 @@
+        // depositNonce (uint64) and configUpdateNonce (uint64)
+        sstore(depositNonce.slot, shr(128, calldataload(164)))
+    }
+}
+```
+
+These L2 contracts are upgraded as part of the Isthmus fork activation block.
+
+## Consensus / Execution software changes
+
+We also add support for these new event versions in the derivation pipeline. For the
+consensus client:
+ - `TransactionDeposited` / `ConfigUpdate` log parsing must support the new version `1`.
+ - If the version is `1`, the log parsing also checks the nonce values are contiguous
+   with the previous nonce seen, read from the `L1Block` contract.
+ - The `L1Block` attributes deposit tx is updated to include the new nonces in the
+   calldata.
+
+Additionally the execution client must be updated, as it reads L1 cost values from the
+first deposit transaction in each block. The 4-byte method signature and calldata
+changes, and the parsing must be updated to support the new format.
+
+The implementation must support both versions `0` and `1` side-by-side. We suggest that
+enforcing version `1` (along with nonces) is implemented in the following hard fork.
+This requires the L1 contract upgrades to be complete.
+
+## Resource Usage
+
+There are two places where resource usage increases:
+1. The gas usage for config updates and deposits increases slightly, because we are now
+   incrementing a nonce for each event emitted.
+2. The calldata size passed to the `L1Block` contract for the first transaction in each
+   L2 block increases by 16 bytes (from 164 bytes to 180 bytes).
+
+We anticipate that derivation, fault proofs, and in the future ZK proofs, will become
+much cheaper as we roll out the ability to skip L1 blocks in the future. There is still
+work to be done to achieve this, as there are other parts of the pipeline that assume
+L1 block references are contiguous.
+
+## Single Point of Failure and Multi Client Considerations
+
+Both consensus and execution client changes are required, so we'll need to schedule
+changes to all relevant clients. Fortunately the required execution changes are very
+minimal.
+
+# Alternatives Considered
+
+There are a number of alternatives considered in the specs
+[issue](https://github.com/ethereum-optimism/specs/issues/330) that inspired the
+implementation. In particular, it is suggested there that no L2 contract changes are
+required. However, without L2 changes, we have to find somewhere else to track the nonce
+values ingested by the L2 in order to enforce the values are contiguous. It made sense
+to piggy back on the existing L1 state tracking in the `L1Block` contract.
+
+# Risks & Uncertainties
+
+One somewhat confusing aspect of this design is the fact that chains that have Isthmus
+activated at genesis will have a `SystemConfig` nonce that starts at `3` (i.e. the
+first event emitted with have a nonce of `4`). This is because the initializer emits
+3 config updates. This initial value will also increase as more config updates are
```

Comment:
I don't think we need to increment the nonce in the L2 genesis, we could just wait for it to be incremented when the L2 picks up the events?